### PR TITLE
Reimplement LDPath transforms as an OSGi service

### PR DESCRIPTION
**Pull request by acoburn**
Resolves: https://jira.duraspace.org/browse/FCREPO-2070

### CODE REVIEW

**Comment by acoburn** [comment]
@awoods: this is my proposal.

File: fcrepo-ldpath/README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    fcrepo.cache.timeout=0
+
+The location of the cached triples
+
+    cache.dir=ldcache
+
+The global timeout for cache entries (in seconds)
+
+    cace.timeout=86400
```
Comment (bseeger) [inline fcrepo-ldpath/README.md]: typo here -- `cace`

**Comment by acoburn** [comment]
@ruebot @dannylamb @diegopino _this_ is what I was talking about the other day. It allows you to start at any Fedora resource URI and walk out the linked data cloud to whatever (arbitrary) depth your LDPath program defines. Unlike `fcrepo-transform` (which is ok for demos but not much beyond that), this will _actually_ be useful for indexing in production repositories; and even more so if your repo links to other, external Fedora repositories.

**Comment by DiegoPino** [comment]
@acoburn wow. I have to digest this to really understand it, but at first look i feel i have to say "it's amazing"

**Comment by ajs6f** [comment]
@acoburn is bringing a hurricane of linked data to Fedora Beach.

**Comment by acoburn** [comment]
Could someone please review/merge this?

**Comment by awoods** [comment]
I will be giving this a review/test shortly.

File: fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+     * @return a configuration for use with an LDClient
+     */
+    public static ClientConfiguration createClient(final Endpoint fedoraEndpoint) {
+        return createClient(null, null, fedoraEndpoint, emptyList(), emptyList());
+    }
+
+    /**
+     * Configure a linked data client suitable for use with a Fedora Repository.
+     * @param fedoraEndpoint a FedoraEndpoint configuration
+     * @param endpoints additional endpoints to enable on the client
```
Comment (awoods) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: Please add the `providers` @param.

Comment (ajs6f) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: We have no current policy on code style, Javadocs, etc. for projects in `fcrepo4-exts`. If we want to create one, we should document it and recommend the use of the same tooling as the core uses. I'm in no way against good Javadocs, but we need to be clear about what the responsibilities of someone putting something in `fcrepo4-exts` are and are not.

Comment (awoods) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: There is a separate issue here of including in the fcrepo-camel-toolbox/pom.xml:
```
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-checkstyle-plugin</artifactId>
      </plugin>
```
But even with the above, apparently we have codified the allowance for omitting @param tags: https://github.com/fcrepo4/fcrepo-build-tools/blob/master/src/main/resources/fcrepo-checkstyle/checkstyle.xml#L69
..which is a little surprising.

File: fcrepo-ldpath/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+       <cm:property name="fcrepo.authUsername" value=""/>
+       <cm:property name="fcrepo.authPassword" value=""/>
+       <cm:property name="fcrepo.authHostname" value="localhost"/>
+       <cm:property name="fcrepo.authPort" value="8080"/>
+       <cm:property name="fcrepo.baseUrl" value="http://localhost:8080/fcrepo/rest"/>
+       <cm:property name="fcrepo.cache.timeout" value="0"/>
+       <cm:property name="rest.prefix" value="/ldpath"/>
+       <cm:property name="rest.host" value="localhost"/>
+       <cm:property name="rest.port" value="9086"/>
+       <cm:property name="cache.dir" value="$[karaf.data]/ldcache"/>
```
Comment (awoods) [inline fcrepo-ldpath/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: When I build `fcrepo-camel-toolbox`, a directory named `$[karaf.data]` is created within the `fcrepo-ldpath` directory... which is undesirable.

File: fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java (hunk 2 above)
Comment (ajs6f) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: I think you have mistaken my meaning. There is no policy under which anyone should include that Maven in an `fcrepo4-exts` project. No one is bound to use the same code conventions that are used in the core. (I certainly wouldn't-- I don't agree with several of them.) If you want to change that policy, I think you need to do it publicly.

Comment (acoburn) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: I would just add that my intention here was to follow the code style of `fcrepo4` and that omitting the param was an oversight on my part.

Comment (ajs6f) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: And I think it's totally fine for those who want to use that style to use it. I have in hand the beginnings of a Fedora reimp (which of course we hope will be one of many). If I have to deal with the same code style that the `fcrepo4` uses, I will not be using `fcrepo4-exts` for it.

File: toolbox-features/src/main/resources/features.xml [hunk 4]
```diff
@@ -81,6 +81,92 @@
...
+    <feature version="${camel.version}">camel-blueprint</feature>
+    <feature version="${camel.version}">camel-jackson</feature>
+    <feature version="${camel.version}">camel-http4</feature>
+    <feature version="${camel.version}">camel-jetty9</feature>
+
+    <bundle dependency="true">mvn:commons-codec/commons-codec/${commons-codec.version}</bundle>
+    <bundle dependency="true">mvn:commons-io/commons-io/${commons-io.version}</bundle>
+
+    <bundle dependency="true">mvn:org.apache.marmotta/ldpath-api/${marmotta.version}</bundle>
+    <!--<bundle dependency="true">mvn:org.apache.marmotta/ldpath-core-bundle/${marmotta.version}</bundle>-->
```
Comment (awoods) [inline toolbox-features/src/main/resources/features.xml]: Is there significance to these commented dependencies?

File: fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+ * @author acoburn
+ */
+public class ClientFactory {
+
+    /**
+     * Configure a linked data client suitable for use with a Fedora Repository.
+     * @param fedoraEndpoint a FedoraEndpoint configuration
+     * @return a configuration for use with an LDClient
+     */
+    public static ClientConfiguration createClient(final Endpoint fedoraEndpoint) {
```
Comment (awoods) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: I am not seeing this method used anywhere. If this is dead code, we should probably remove it.

File: fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        return createClient(null, null, fedoraEndpoint, emptyList(), emptyList());
+    }
+
+    /**
+     * Configure a linked data client suitable for use with a Fedora Repository.
+     * @param fedoraEndpoint a FedoraEndpoint configuration
+     * @param endpoints additional endpoints to enable on the client
+     * @return a configuration for use with an LDClient
+     */
+    public static ClientConfiguration createClient(final Endpoint fedoraEndpoint, final List<Endpoint> endpoints,
```
Comment (awoods) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/ClientFactory.java]: I am not seeing this method used anywhere. If this is dead code, we should probably remove it.

File: fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/LDPathRouter.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    public static final String FEDORA_URI = "CamelFedoraUri";
+
+    /**
+     * Configure the message route workflow.
+     */
+    public void configure() throws Exception {
+
+        /**
+         * Expose a RESTful endpoint for re-indexing
```
Comment (awoods) [inline fcrepo-ldpath/src/main/java/org/fcrepo/camel/ldpath/LDPathRouter.java]: Not `re-indexing`.

File: fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+public class RouteIT extends CamelBlueprintTestSupport {
+
+    private static final Logger LOGGER = getLogger(RouteIT.class);
+
+    private final ObjectMapper MAPPER = new ObjectMapper();
+
+    private final CloseableHttpClient httpclient = HttpClients.createDefault();
+
+    @EndpointInject(uri = "mock:result")
+    protected MockEndpoint resultEndpoint;
```
Comment (awoods) [inline fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java]: dead code

File: fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+    @Produce(uri = "direct:reindex")
+    protected ProducerTemplate template;
```
Comment (awoods) [inline fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java]: dead code

File: fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+    private String fullPath = "";
```
Comment (awoods) [inline fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java]: dead code

File: fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        try {
+            final HttpResponse response = httpclient.execute(new HttpGet(url));
+            return EntityUtils.toString(response.getEntity(), "UTF-8");
+        } catch (IOException ex) {
+            LOGGER.debug("Unable to extract HttpEntity response into an InputStream: ", ex);
+            return "";
+        }
+    }
+
+    private String post(final String url) {
```
Comment (awoods) [inline fcrepo-ldpath/src/test/java/org/fcrepo/camel/ldpath/integration/RouteIT.java]: dead code

File: fcrepo-ldpath/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 3 above)
Comment (acoburn) [inline fcrepo-ldpath/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I completely agree. Adding a system property to maven helps avoid this.

**Pull request merged**